File: src/core/geometry.ts
/**
 * Geometry utilities for circular diagram calculations
 *
 * @license MIT
 * @copyright 2026 CGA Management Ltd
 * @see https://www.cgamanagement.co.uk/category/tools
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export interface Point {
  x: number;
  y: number;
}

/**
 * Convert polar coordinates to Cartesian (SVG) coordinates
 * @param cx Center X
 * @param cy Center Y
 * @param radius Distance from center
 * @param angleDegrees Angle in degrees (0 = right, 90 = down in SVG coords)
 */
export function polarToCartesian(
  cx: number,
  cy: number,
  radius: number,
  angleDegrees: number
): Point {
  const angleRadians = (angleDegrees * Math.PI) / 180;
  return {
    x: cx + radius * Math.cos(angleRadians),
    y: cy + radius * Math.sin(angleRadians),
  };
}

/**
 * Generate SVG arc path data
 * @param cx Center X
 * @param cy Center Y
 * @param radius Arc radius
 * @param startAngle Start angle in degrees
 * @param endAngle End angle in degrees
 * @returns SVG path d attribute for the arc
 */
export function describeArc(
  cx: number,
  cy: number,
  radius: number,
  startAngle: number,
  endAngle: number
): string {
  const start = polarToCartesian(cx, cy, radius, startAngle);
  const end = polarToCartesian(cx, cy, radius, endAngle);

  // Determine if arc should be drawn the "long way" around
  const angleDiff = endAngle - startAngle;
  const largeArcFlag = Math.abs(angleDiff) > 180 ? 1 : 0;

  // Sweep direction (1 = clockwise in SVG)
  const sweepFlag = angleDiff > 0 ? 1 : 0;

  return `M ${start.x} ${start.y} A ${radius} ${radius} 0 ${largeArcFlag} ${sweepFlag} ${end.x} ${end.y}`;
}

/**
 * Generate SVG path for a filled segment (pie slice or ring segment)
 * @param cx Center X
 * @param cy Center Y
 * @param innerRadius Inner radius (0 for full pie slice)
 * @param outerRadius Outer radius
 * @param startAngle Start angle in degrees
 * @param endAngle End angle in degrees
 * @returns SVG path d attribute for the filled segment
 */
export function segmentPath(
  cx: number,
  cy: number,
  innerRadius: number,
  outerRadius: number,
  startAngle: number,
  endAngle: number
): string {
  const outerStart = polarToCartesian(cx, cy, outerRadius, startAngle);
  const outerEnd = polarToCartesian(cx, cy, outerRadius, endAngle);
  const innerStart = polarToCartesian(cx, cy, innerRadius, startAngle);
  const innerEnd = polarToCartesian(cx, cy, innerRadius, endAngle);

  const angleDiff = endAngle - startAngle;
  const largeArcFlag = Math.abs(angleDiff) > 180 ? 1 : 0;

  if (innerRadius === 0) {
    // Full pie slice from center
    return [
      `M ${cx} ${cy}`,
      `L ${outerStart.x} ${outerStart.y}`,
      `A ${outerRadius} ${outerRadius} 0 ${largeArcFlag} 1 ${outerEnd.x} ${outerEnd.y}`,
      'Z',
    ].join(' ');
  }

  // Ring segment (donut slice)
  return [
    `M ${outerStart.x} ${outerStart.y}`,
    `A ${outerRadius} ${outerRadius} 0 ${largeArcFlag} 1 ${outerEnd.x} ${outerEnd.y}`,
    `L ${innerEnd.x} ${innerEnd.y}`,
    `A ${innerRadius} ${innerRadius} 0 ${largeArcFlag} 0 ${innerStart.x} ${innerStart.y}`,
    'Z',
  ].join(' ');
}

/**
 * Generate SVG path for a filled arc segment within a facet
 * Used for showing scores as filled arcs
 * @param cx Center X
 * @param cy Center Y
 * @param innerRadius Inner radius (center hub edge)
 * @param outerRadius Outer radius (based on score)
 * @param startAngle Start angle of facet
 * @param endAngle End angle of facet
 */
export function facetScorePath(
  cx: number,
  cy: number,
  innerRadius: number,
  outerRadius: number,
  startAngle: number,
  endAngle: number
): string {
  return segmentPath(cx, cy, innerRadius, outerRadius, startAngle, endAngle);
}

/**
 * Calculate the angle span for each segment
 * @param segmentCount Number of segments (must be > 0)
 * @returns Angle in degrees per segment
 * @throws Error if segmentCount is not greater than 0
 */
export function segmentAngle(segmentCount: number): number {
  if (segmentCount <= 0) {
    throw new Error('segmentCount must be greater than 0');
  }
  return 360 / segmentCount;
}

/**
 * Calculate angles for facets within a segment
 * @param segmentStartAngle Start angle of the segment
 * @param segmentEndAngle End angle of the segment
 * @param facetCount Number of facets in segment (must be > 0)
 * @returns Array of { startAngle, endAngle, midAngle } for each facet
 * @throws Error if facetCount is not greater than 0
 */
export function facetAngles(
  segmentStartAngle: number,
  segmentEndAngle: number,
  facetCount: number
): Array<{ startAngle: number; endAngle: number; midAngle: number }> {
  if (facetCount <= 0) {
    throw new Error('facetCount must be greater than 0');
  }
  const totalAngle = segmentEndAngle - segmentStartAngle;
  const facetAngleSpan = totalAngle / facetCount;

  return Array.from({ length: facetCount }, (_, i) => {
    const startAngle = segmentStartAngle + i * facetAngleSpan;
    const endAngle = startAngle + facetAngleSpan;
    const midAngle = (startAngle + endAngle) / 2;
    return { startAngle, endAngle, midAngle };
  });
}

/**
 * Calculate radius for a given score level
 * Uses level-based calculation where score 1 = 1/5, score 5 = 5/5 (for 1-5 scale)
 * @param score The score value (will be clamped to valid range)
 * @param minScore Minimum score in scale
 * @param maxScore Maximum score in scale
 * @param innerRadius Radius of center hub
 * @param outerRadius Maximum outer radius
 * @throws Error if minScore > maxScore or innerRadius >= outerRadius
 */
export function scoreToRadius(
  score: number,
  minScore: number,
  maxScore: number,
  innerRadius: number,
  outerRadius: number
): number {
  if (minScore > maxScore) {
    throw new Error('minScore must be less than or equal to maxScore');
  }
  if (innerRadius >= outerRadius) {
    throw new Error('innerRadius must be less than outerRadius');
  }
  // When min === max, return full outer radius (flat diagram)
  if (minScore === maxScore) {
    return outerRadius;
  }
  // Clamp score to valid range
  const clampedScore = Math.max(minScore, Math.min(maxScore, score));
  const levels = maxScore - minScore + 1; // 5 levels for 1-5 scale
  const radiusRange = outerRadius - innerRadius;
  const normalizedScore = (clampedScore - minScore + 1) / levels;
  return innerRadius + normalizedScore * radiusRange;
}

/**
 * Calculate radii for ring grid lines
 * @param ringCount Number of rings (must be > 0)
 * @param innerRadius Inner radius (center hub)
 * @param outerRadius Outer radius
 * @throws Error if ringCount is not greater than 0 or innerRadius >= outerRadius
 */
export function ringRadii(
  ringCount: number,
  innerRadius: number,
  outerRadius: number
): number[] {
  if (ringCount <= 0) {
    throw new Error('ringCount must be greater than 0');
  }
  if (innerRadius >= outerRadius) {
    throw new Error('innerRadius must be less than outerRadius');
  }
  const step = (outerRadius - innerRadius) / ringCount;
  return Array.from({ length: ringCount + 1 }, (_, i) => innerRadius + i * step);
}

/**
 * Calculate text anchor and rotation for labels at an angle
 * @param angle Angle in degrees
 * @returns Object with textAnchor and rotation for readable text
 */
export function labelOrientation(angle: number): {
  textAnchor: 'start' | 'middle' | 'end';
  rotation: number;
  alignmentBaseline: string;
} {
  // Normalize angle to 0-360
  const normalizedAngle = ((angle % 360) + 360) % 360;

  // Text should be readable (not upside down)
  // Right side (315-45): start anchor
  // Bottom (45-135): middle anchor
  // Left side (135-225): end anchor
  // Top (225-315): middle anchor

  let textAnchor: 'start' | 'middle' | 'end' = 'middle';
  let rotation = normalizedAngle + 90; // Tangent to circle
  let alignmentBaseline = 'middle';

  // Flip text that would be upside down
  if (normalizedAngle > 90 && normalizedAngle < 270) {
    rotation += 180;
  }

  // Adjust anchor based on position
  if (normalizedAngle >= 350 || normalizedAngle <= 10) {
    textAnchor = 'start';
    alignmentBaseline = 'middle';
  } else if (normalizedAngle >= 170 && normalizedAngle <= 190) {
    textAnchor = 'end';
    alignmentBaseline = 'middle';
  }

  return { textAnchor, rotation, alignmentBaseline };
}

File: python/radial_diagram/geometry.py
"""
Geometry utilities for circular diagram calculations.

1:1 port of src/core/geometry.ts. All functions are pure.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Literal


@dataclass
class Point:
    x: float
    y: float


@dataclass
class FacetAngle:
    startAngle: float
    endAngle: float
    midAngle: float


@dataclass
class LabelOrientation:
    textAnchor: Literal["start", "middle", "end"]
    rotation: float
    alignmentBaseline: str


def polar_to_cartesian(
    cx: float, cy: float, radius: float, angle_degrees: float
) -> Point:
    angle_radians = (angle_degrees * math.pi) / 180
    return Point(
        x=cx + radius * math.cos(angle_radians),
        y=cy + radius * math.sin(angle_radians),
    )


def describe_arc(
    cx: float,
    cy: float,
    radius: float,
    start_angle: float,
    end_angle: float,
) -> str:
    start = polar_to_cartesian(cx, cy, radius, start_angle)
    end = polar_to_cartesian(cx, cy, radius, end_angle)

    angle_diff = end_angle - start_angle
    large_arc_flag = 1 if abs(angle_diff) > 180 else 0
    sweep_flag = 1 if angle_diff > 0 else 0

    return (
        f"M {start.x} {start.y} A {radius} {radius} 0 "
        f"{large_arc_flag} {sweep_flag} {end.x} {end.y}"
    )


def segment_path(
    cx: float,
    cy: float,
    inner_radius: float,
    outer_radius: float,
    start_angle: float,
    end_angle: float,
) -> str:
    outer_start = polar_to_cartesian(cx, cy, outer_radius, start_angle)
    outer_end = polar_to_cartesian(cx, cy, outer_radius, end_angle)
    inner_start = polar_to_cartesian(cx, cy, inner_radius, start_angle)
    inner_end = polar_to_cartesian(cx, cy, inner_radius, end_angle)

    angle_diff = end_angle - start_angle
    large_arc_flag = 1 if abs(angle_diff) > 180 else 0

    if inner_radius == 0:
        return " ".join(
            [
                f"M {cx} {cy}",
                f"L {outer_start.x} {outer_start.y}",
                f"A {outer_radius} {outer_radius} 0 {large_arc_flag} 1 "
                f"{outer_end.x} {outer_end.y}",
                "Z",
            ]
        )

    return " ".join(
        [
            f"M {outer_start.x} {outer_start.y}",
            f"A {outer_radius} {outer_radius} 0 {large_arc_flag} 1 "
            f"{outer_end.x} {outer_end.y}",
            f"L {inner_end.x} {inner_end.y}",
            f"A {inner_radius} {inner_radius} 0 {large_arc_flag} 0 "
            f"{inner_start.x} {inner_start.y}",
            "Z",
        ]
    )


def facet_score_path(
    cx: float,
    cy: float,
    inner_radius: float,
    outer_radius: float,
    start_angle: float,
    end_angle: float,
) -> str:
    return segment_path(cx, cy, inner_radius, outer_radius, start_angle, end_angle)


def segment_angle(segment_count: int) -> float:
    if segment_count <= 0:
        raise ValueError("segmentCount must be greater than 0")
    return 360 / segment_count


def facet_angles(
    segment_start_angle: float,
    segment_end_angle: float,
    facet_count: int,
) -> list[FacetAngle]:
    if facet_count <= 0:
        raise ValueError("facetCount must be greater than 0")
    total_angle = segment_end_angle - segment_start_angle
    facet_angle_span = total_angle / facet_count

    result: list[FacetAngle] = []
    for i in range(facet_count):
        start_angle = segment_start_angle + i * facet_angle_span
        end_angle = start_angle + facet_angle_span
        mid_angle = (start_angle + end_angle) / 2
        result.append(
            FacetAngle(startAngle=start_angle, endAngle=end_angle, midAngle=mid_angle)
        )
    return result


def score_to_radius(
    score: float,
    min_score: float,
    max_score: float,
    inner_radius: float,
    outer_radius: float,
) -> float:
    if min_score > max_score:
        raise ValueError("minScore must be less than or equal to maxScore")
    if inner_radius >= outer_radius:
        raise ValueError("innerRadius must be less than outerRadius")
    if min_score == max_score:
        return outer_radius
    clamped_score = max(min_score, min(max_score, score))
    levels = max_score - min_score + 1
    radius_range = outer_radius - inner_radius
    normalized_score = (clamped_score - min_score + 1) / levels
    return inner_radius + normalized_score * radius_range


def ring_radii(
    ring_count: int, inner_radius: float, outer_radius: float
) -> list[float]:
    if ring_count <= 0:
        raise ValueError("ringCount must be greater than 0")
    if inner_radius >= outer_radius:
        raise ValueError("innerRadius must be less than outerRadius")
    step = (outer_radius - inner_radius) / ring_count
    return [inner_radius + i * step for i in range(ring_count + 1)]


def label_orientation(angle: float) -> LabelOrientation:
    normalized_angle = ((angle % 360) + 360) % 360

    text_anchor: Literal["start", "middle", "end"] = "middle"
    rotation = normalized_angle + 90
    alignment_baseline = "middle"

    if 90 < normalized_angle < 270:
        rotation += 180

    if normalized_angle >= 350 or normalized_angle <= 10:
        text_anchor = "start"
        alignment_baseline = "middle"
    elif 170 <= normalized_angle <= 190:
        text_anchor = "end"
        alignment_baseline = "middle"

    return LabelOrientation(
        textAnchor=text_anchor,
        rotation=rotation,
        alignmentBaseline=alignment_baseline,
    )
